text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,755 @@
+        }
+
+        Ok(())
+    }
+}
+```
+
+Now we have an error reporter that is ready for use, a simple program using it
+would look like this.
+
+```rust
+fn main() -> Result<(), ErrorReporter> {
+    let path = "./path/to/instrs.json";
+    let _instrs = read_instrs(path.into())?;
+}
+```
+
+Which, if run without creating the `instrs.json` file prints this error report:
+
+```
+Error:
+    0: Failed to read instrs from ./path/to/instrs.json
+        at instrs.rs:42
+    1: No such file or directory (os error 2)
+```
+
+Mission accomplished! The error trait gave us everything we needed to build
+error reports enriched by context relevant to our application. This same
+pattern can be implement many error reporting patterns, such as including help
+text, spans, http status codes, or backtraces in errors which are still
+accessible after the error has been converted to a `dyn Error`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following changes need to be made to implement this proposal:
+
+### Add a `Request` type to `libcore` for type-indexed context
+
+`Request` is a type to emulate nested dynamic typing. This type fills the same
+role as `&dyn Any` except that it supports other trait objects as the requested
+type.
+
+Here is the implementation for the proof of concept, based on Nika Layzell's
+[object-provider crate]:
+
+A usable version of this is available in the [proof of concept] repo under
+`fakecore/src/any.rs`.
+
+```rust
+use core::any::TypeId;
+use core::marker::PhantomData;
+
+mod private {
+    pub trait Response<'a>: 'a {}
+}
+
+/// A response to a ref request.
+struct RefResponse<'a, T: ?Sized + 'static>(Option<&'a T>);
+impl<'a, T: ?Sized + 'static> private::Response<'a> for RefResponse<'a, T> {}
+
+/// A response to a value request.
+struct ValueResponse<T: 'static>(Option<T>);
+impl<'a, T: 'static> private::Response<'a> for ValueResponse<T> {}
+
+/// A dynamic request for an object based on its type.
+pub struct Request<'a, R = dyn private::Response<'a>>
+where
+    R: ?Sized + private::Response<'a>,
+{
+    marker: PhantomData<&'a ()>,
+
+    /// A `TypeId` marker for the type stored in `R`.
+    ///
+    /// Will be the TypeId of either `RefResponse<'static, T>` or
+    /// `ValueResponse<T>`.
+    type_id: TypeId,
+
+    /// A type erased `RefResponse` or `ValueResponse` containing the response
+    /// value.
+    response: R,
+}
+
+impl<'a> Request<'a> {
+    /// Perform a checked downcast of `response` to `Option<&'a T>`
+    fn downcast_ref_response<'b, T: ?Sized + 'static>(
+        &'b mut self,
+    ) -> Option<&'b mut RefResponse<'a, T>> {
+        if self.is_ref::<T>() {
+            // safety: If `self.is_ref::<T>()` returns true, `response` must be
+            // of the correct type. This is enforced by the private `type_id`
+            // field.
+            Some(unsafe { &mut *(&mut self.response as *mut _ as *mut RefResponse<'a, T>) })
```

Comment:
Maybe my coercion-fu is lacking, but what is the `as *mut _` part in the middle achieving?

Reply:
Mutable references (in this case `&mut R`) can only be cast to pointers to the same type. Here, that pointer is then cast from `*mut R` to `*mut RefResponse<'a, T>`.

Reply:
All of this is out of date anyways, I need to update it to use https://docs.rs/dyno/ as the example